crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,411 @@
+        request: ReadRowsRequest,
+    ) -> Result<Vec<(Vec<u8>, Vec<(Vec<u8>, Vec<u8>)>)>> {
+        let mut result = vec![];
+        let mut response = self.client.read_rows(request).await?.into_inner();
+
+        let mut row_key = None;
+        let mut row = vec![];
+        let mut cell_value = vec![];
+        let mut cell_name = None;
+        let mut timestamp = 0;
+
+        while let Some(message) = response.message().await? {
+            for mut chunk in message.chunks.into_iter() {
+                // new row check
+                if !chunk.row_key.is_empty() {
+                    row_key = Some(chunk.row_key);
+                }
+                match chunk.qualifier {
+                    // new cell started
+                    Some(qualifier) => {
+                        if let Some(cell_name) = cell_name {
+                            row.push((cell_name, cell_value));
+                            cell_value = vec![];
+                        }
+                        cell_name = Some(qualifier);
+                        timestamp = chunk.timestamp_micros;
+                        cell_value.append(&mut chunk.value);
+                    }
+                    None => {
+                        if chunk.timestamp_micros == 0 {
+                            cell_value.append(&mut chunk.value);
+                        } else if chunk.timestamp_micros >= timestamp {
+                            // newer version of cell is available
+                            timestamp = chunk.timestamp_micros;
+                            cell_value = chunk.value;
+                        }
+                    }
+                }
+                if chunk.row_status.is_some() {
+                    if let Some(RowStatus::CommitRow(_)) = chunk.row_status {
+                        if let Some(cell_name) = cell_name {
+                            row.push((cell_name, cell_value));
+                        }
+                        if let Some(row_key) = row_key {
+                            result.push((row_key, row));
+                        }
+                    }
+                    row_key = None;
+                    row = vec![];
+                    cell_value = vec![];
+                    cell_name = None;
+                }
+            }
+        }
+        Ok(result)
+    }
+
+    async fn multi_set(
+        &mut self,
+        table_name: &str,
+        values: impl IntoIterator<Item = (Bytes, Vec<(&str, Bytes)>)> + std::marker::Send,
+    ) -> Result<()> {
+        let mut entries = vec![];
+        for (row_key, cells) in values {
+            let mutations = cells
+                .into_iter()
+                .map(|(column_name, value)| Mutation {
+                    mutation: Some(mutation::Mutation::SetCell(SetCell {
+                        family_name: COLUMN_FAMILY_NAME.to_string(),
+                        column_qualifier: column_name.to_owned().into_bytes(),
+                        timestamp_micros: -1,
```

Comment:
Do we want to try and have this timestamp be accurate and include the timestamp of the checkpoint where the data resides?

Reply:
That's an internal timestamp used by Bigtable to determine which cell version is the latest. The value -1 is a special one that instructs Bigtable to rely on server time

Reply:
oh neat!

Reply:
would you mind maybe providing a comment for that so that?